### ISSUE TITLE
hard requirement on 5.8.8?

### ISSUE THREAD

**Issue by bnjf** [RexOps/Rex#746]
Is there a particular reason for the hard requirement on 5.8._8_?  Some older OSs (Solaris 10 in particular) ship with an older version, with vendor patches.  I've changed it to 5.8.4 and rex seems to connect fine, run its jobs, etc.

``` patch

---
 Makefile.PL | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile.PL b/Makefile.PL
index f72c99b..d824a75 100644
--- a/Makefile.PL
+++ b/Makefile.PL
@@ -2,7 +2,7 @@
 use strict;
 use warnings;

-use 5.008008;
+use 5.008004;

 use ExtUtils::MakeMaker;

@@ -21,7 +21,7 @@ my %WriteMakefileArgs = (
     "bin/rex"
   ],
   "LICENSE" => "apache",
-  "MIN_PERL_VERSION" => "5.008008",
+  "MIN_PERL_VERSION" => "5.008004",
   "NAME" => "Rex",
   "PREREQ_PM" => {
     "Carp" => 0,
-- 
2.3.2
```

**Comment by ferki** [comment]
@bnjf: hi, in general the dependency we aim for is "the latest patch versoin for every minor version since 5.8". So for 5.8.x it is 5.8.9.

We decreased it to 5.8.8 only because of CentOS/Red Hat 5, which is technically shipped with 5.8.8, but has a backported fix to this issue (for details, see also #578). Of course, a better solution for this would be to accept 5.8.8 only on CentOS/Red Hat 5.

We might be able to do the same exception for Solaris 10, but it doesn't seem that easy to lower this dependency to 5.8.4 _in  general_ (even though according to `Test::MinimumVersion`, the code syntax and langauge features rex uses, is compatible).

Unfortunately I'm not overly familiar with Solaris and its bugs/documentation - can you point us to some kind of "perl patching history" to see if Solaris' version of perl-5.8.4 has at least the same patch as mentioned above?

The other solution is to replace all regular expressions in the code to work without the `/m` modifier - which is tedious, but probably the proper way to go if we start adding more and more exceptions for old OSes.

Just for the sake of completeness, there's also another factor here, that we actually planned to drop 5.8.x support once CentOS/Red Hat 5 goes EOL, but looks like Solaris 10 will be supported until 2021, which would mean we should support a 11 years old perl version for 6 more years.

So I guess we shuold find a nice compromise somewhere in between all those factors - which I'm pretty sure we can do :)

ps.: also, we could use Solaris KVM images for automated testing on build.rexify.org - if you can contribute one, please open an issue in the [rex-build](https://github.com/RexOps/rex-build) repo about it (or ping us via IRC: #rex at Freenode).

**Comment by bnjf** [comment]
(I'll see what I can do about a KVM image)

No good, it's hit by that bug.  I'll raise a case with Oracle.

```
$ perl -wle '$a = "begin\nend"; $a =~ /^(begin)$/m; print $1' 
begin
$ perl -wle '$a = "begin\nend"; $a =~ qr/^(begin)$/m; print $1'
Use of uninitialized value in print at -e line 1.

$ 
```

For reference, here's the applied patches:
  Locally applied patches:
        22667 The optree builder was looping when constructing the ops ...
        22715 Upgrade to FileCache 1.04
        22733 Missing copyright in the README.
        22746 fix a coredump caused by rv2gv not fully converting a PV ...
        22755 Fix 29149 - another UTF8 cache bug hit by substr.
        22774 [perl #28938] split could leave an array without ...
        22775 [perl #29127] scalar delete of empty slice returned garbage
        22776 [perl #28986] perl -e "open m" crashes Perl
        22777 add test for change #22776 ("open m" crashes Perl)
        22778 add test for change #22746 ([perl #29102] Crash on assign ...
        22781 [perl #29340] Bizarre copy of ARRAY make sure a pad op's ...
        22796 [perl #29346] Double warning for int(undef) and abs(undef) ...
        22818 BOM-marked and (BOMless) UTF-16 scripts not working
        22823 [perl #29581] glob() misses a lot of matches
        22827 Smoke [5.9.2] 22818 FAIL(F) MSWin32 WinXP/.Net SP1 (x86/1 cpu)
        22830 [perl #29637] Thread creation time is hypersensitive
        22831 improve hashing algorithm for ptr tables in perl_clone: ...
        22839 [perl #29790] Optimization busted: '@a = "b", sort @a' ...
        22850 [PATCH] 'perl -v' fails if local_patches contains code snippets
        22852 TEST needs to ignore SCM files
        22886 Pod::Find should ignore SCM files and dirs
        22888 Remove redundant %SIG assignments from FileCache
        23006 [perl #30509] use encoding and "eq" cause memory leak
        23074 Segfault using HTML::Entities
        23106 Numeric comparison operators mustn't compare addresses of ...
        23320 [perl #30066] Memory leak in nested shared data structures ...
        23321 [perl #31459] Bug in read()
        27722 perlio.c breaks on Solaris/gcc when > 256 FDs are available
        SPRINTF0 - fixes for sprintf formatting issues - CVE-2005-3962
        6663288 Upgrade to CGI.pm 3.33
        REGEXP0 - fix for UTF-8 recoding in regexps - CVE-2007-5116
        6758953 Perl Sys::Syslog can log messages with wrong severity
        6935710 patch 141552-01 causes an error when the Syslog perl module is used
        6961567 CVE-2010-1168 - Safe.pm 2.24 and earlier may allow attackers to break out of safe compartment
        7004391 CGI.pm needs update to filter newlines in headers (CVE-2010-2761, CVE-2010-4411)
        7116936 Problem with utility/perl
        7125218 Problem with utility/perl
        16427794 problem in UTILITY/PERL
        16590139 problem in UTILITY/PERL
        16894348 problem in UTILITY/PERL
        15808423 problem in UTILITY/PERL
        17185829 problem in UTILITY/PERL
        15290783 SUNBT6333830 Perl should use setenv(3c) and unsetenv(3c) calls

**Comment by kablamo** [comment]
Maybe build your own perl?  http://perlbrew.pl/ or https://github.com/tokuhirom/Perl-Build

**Comment by ferki** [comment]
We can (should :) also mention this on the website, and/or set up an "instructions for solaris" page, or something along those lines.

**Comment by bnjf** [comment]
hang in there, I've got a PR open with Oracle... might be able to get the problem with multiline qr's patched

**Comment by ferki** [comment]
## TL;DR

I’m on the opinion that moving forward with the minimum required Perl version to run Rex is the responsible way. I recommend reading the [leaving perl v5.8 behind](https://rjbs.cloud/blog/2023/01/leaving-perl-v5.8-behind/) post for a good collection of reasons. Rex now requires at least Perl 5.12.5.

## Longer version

Let me react to some of the main points from this issue before closing it:

### Perl 5.8.4 and Solaris 10

> [...] hard requirement on 5.8.8? Some older OSs (Solaris 10 in particular) ship with an older version, with vendor patches. I’ve changed it to 5.8.4 [...]

> [...] which would mean we should support a 11 years old perl version for 6 more years.

At the time of opening it meant 5.8.9, with an exception for 5.8.8 on CentOS/RHEL (which was so heavily patched that it could be considered 5.8.9 for the reasons of Rex in practice).

The release dates, and the age of various Perl versions look like this:

| version | release date | age at opening | age today |
|:-------:|:------------:|---------------:|:---------:|
|  5.8.4  |  2004-04-21  |    ~11 years   | ~19 years |
|  5.8.8  |  2006-01-31  |     ~9 years   | ~17 years |
|  5.8.9  |  2008-12-14  |     ~7 years   | ~15 years |

Later, the minimum requirement has changed to 5.10.1, then 5.12.5. A similar table shows details at these changes:

| version | release date | required since | age at change | latest at change |
|--------:|:------------:|:--------------:|--------------:|:-----------------:|
|  5.8.0  |  2002-07-18  |   2011-04-01   |    ~9 years   |      5.14.2      |
|  5.8.9  |  2008-12-14  |   2014-11-09   |    ~6 years   |      5.20.1      |
|  5.10.1 |  2009-08-22  |   2019-08-17   |   ~10 years   |      5.30.2      |
|  5.12.5 |  2012-11-10  |   2023-01-30   |   ~10 years   |      5.36.0      |

It is likely that even newer Perl versions will be required in the future, and/or the grace period gap will be reduced further. The [Supported Perl versions](https://github.com/RexOps/Rex/blob/92c1a37b2cc08e29992490f6607f2351fbeb2a97/CONTRIBUTING.md#supported-perl-versions) section of the Contributing guide now describes the general approach as:

> Rex aims to run even on older Perl versions up to 10 years old.
> [...]
> On top of the supported minimum version of Perl, the goal is to support the latest versions.

As a similar rule in general about [Cross platform support](https://github.com/RexOps/Rex/blob/92c1a37b2cc08e29992490f6607f2351fbeb2a97/CONTRIBUTING.md#cross-platform-support) and [Supported OSes](https://github.com/RexOps/Rex/blob/92c1a37b2cc08e29992490f6607f2351fbeb2a97/CONTRIBUTING.md#supported-oses):

> running Rex [or] managing an endpoint with Rex is only going to be supported on platforms which are actively maintained by their respective upstream teams.

The standard Premier Support of Solaris 10 has ended on 2018-01-31 (paid extended support was available until 2021-01-31 and later extended to 2024-01-31).

As a completely volunteer-based open source project, we currently have no means to keep supporting Perl 5.8.4 or Solaris 10, and I prefer encouraging keeping systems updated instead of postponing the inevitable end-of-support.

Various service providers might be able to provide commercial support for Rex on old Perl and operating system versions (full disclosure: including myself).

> We can (should :) also mention this on the website, and/or set up an “instructions for solaris” page, or something along those lines.

Installing Rex from CPAN into a separate Perl environment is the recommended way to get Rex since 2019-07-02 (see also RexOps/rexify-website#116).

> hang in there, I’ve got a PR open with Oracle... might be able to get the problem with multiline qr’s patched

@bjnf: I would still be curious what happened to that side of the story, please share if you can.

**Comment by bnjf** [comment]
I'd forgotten about this almost completely.  I no longer work with Solaris, nor do I have access to Oracle's support portal, but given Solaris 10 was on life support at the time (although I'm not surprised about the extension to 2024-01-31) I anticipate the response would've been "Solaris 11 is available and has a more recent Perl".

Thanks for getting back to me, @ferki!

**Comment by ferki** [comment]
@bnjf: thanks for the patience, the updates, and bringing this topic up back then! Despite the lack of activity on this issue, it helped shape how I think about what Rex can sustainably support.

And yes, Solaris 11 has newer Perl versions out of the box, and we even get [test results from it via CPAN Testers](http://fast-matrix.cpantesters.org/?dist=Rex) for us (thanks! :heart:)
